### ISSUE TITLE
VMSS assigned identities are not preserved after performing node pool update

### ISSUE THREAD

**Issue by atiasadir** [Azure/aks-engine#4013]
We observed that VMSS assigned identities are not preserved after performing node pool update via aks-engine v0.57.0

**Steps To Reproduce**

1. **create a cluster**
Make sure it has useManagedIdentity: true
**Deploy command**
`aks-engine deploy --azure-env AzurePublicCloud --subscription-id <subscripttion>--resource-group resource-group-test --location centralus --api-model aks-engine-template.json --client-id <client id> --client-secret <secret> --output-directory aksoutput `

here is the attached api-model:
```{
    "apiVersion": "vlabs",
    "properties": {
        "orchestratorProfile": {
            "orchestratorType": "Kubernetes",
            "orchestratorVersion": "1.16.15",
            "kubernetesConfig": {
                "loadBalancerSku": "standard",
                "apiServerConfig": {
                    "--enable-admission-plugins": "NamespaceLifecycle,LimitRanger,ServiceAccount,DefaultStorageClass,DefaultTolerationSeconds,MutatingAdmissionWebhook,ValidatingAdmissionWebhook,ResourceQuota,ExtendedResourceToleration",
                    "--tls-cipher-suites": "",
                    "--oidc-client-id": "XXX",
                    "--oidc-groups-claim": "roles",
                    "--oidc-issuer-url": "https://login.microsoftonline.com/XXX/v2.0",
                    "--oidc-username-claim": "email",
                    "--oidc-username-prefix": "oidc:"
                },
                "useManagedIdentity": true,
                "userAssignedID": "XXX",
                "addons": [
                    {
                        "name": "cluster-autoscaler",
                        "enabled": true,
                        "pools": [
                            {
                                "name": "atpdev9cus",
                                "config": {
                                    "min-nodes": "2",
                                    "max-nodes": "99"
                                }
                            }
                        ],
                        "containers": [
                            {
                                "name": "cluster-autoscaler",
                                "image": "mcr.microsoft.com/oss/kubernetes/autoscaler/cluster-autoscaler:v1.16.4",
                                "cpuRequests": "100m",
                                "memoryRequests": "2Gi",
                                "cpuLimits": "100m",
                                "memoryLimits": "2Gi"
                            }
                        ],
                        "config": {
                            "scan-interval": "30s"
                        }
                    },
                    {
                        "name": "tiller",
                        "enabled": false
                    },
                    {
                        "name": "keyvault-flexvolume",
                        "enabled": false
                    },
                    {
                        "name": "blobfuse-flexvolume",
                        "enabled": false
                    },
                    {
                        "name": "kubernetes-dashboard",
                        "enabled": false
                    },
                    {
                        "name": "csi-secrets-store",
                        "enabled": false
                    },
                    {
                        "name": "ip-masq-agent",
                        "enabled": true,
                        "containers": [
                            {
                                "name": "ip-masq-agent",
                                "image": "mcr.microsoft.com/oss/kubernetes/ip-masq-agent:v2.5.0",
                                "cpuRequests": "50m",
                                "memoryRequests": "50Mi",
                                "cpuLimits": "50m",
                                "memoryLimits": "250Mi"
                            }
                        ],
                        "config": {
                            "non-masq-cni-cidr": "168.63.129.16/32",
                            "non-masquerade-cidr": "10.10.0.0/16"
                        }
                    }
                ]
            }
        },
        "masterProfile": {
            "count": 3,
            "dnsPrefix": "resource-group-test",
            "vmSize": "Standard_D8s_v3",
            "osDiskSizeGB": 64,
            "vnetSubnetId": "/subscriptions/XXX/resourceGroups/resource-group-test/providers/Microsoft.Network/virtualNetworks/clusterVnet/subnets/masterSubnet",
            "firstConsecutiveStaticIP": "10.10.0.230",
            "vnetCidr": "10.10.0.0/16",
            "subjectAltNames": [
                "resource-group-test.wdatp.windows.com"
            ]
        },
        "agentPoolProfiles": [
            {
                "name": "atpdev9cus",
                "count": 6,
                "vmSize": "Standard_D4_v2",
                "vnetSubnetId": "/subscriptions/XXX/resourceGroups/resource-group-test/providers/Microsoft.Network/virtualNetworks/clusterVnet/subnets/agentSubnet",
                "availabilityProfile": "VirtualMachineScaleSets",
                "scaleSetPriority": "Regular",
                "storageProfile": "ManagedDisks",
                "osDiskSizeGB": 400,
                "acceleratedNetworkingEnabled": true,
                "singlePlacementGroup": false
            },
            {
                "name": "atp2dev9cus",
                "count": 6,
                "vmSize": "Standard_D32s_v3",
                "vnetSubnetId": "/subscriptions/XXX/resourceGroups/resource-group-test/providers/Microsoft.Network/virtualNetworks/clusterVnet/subnets/agent1Subnet",
                "availabilityProfile": "VirtualMachineScaleSets",
                "scaleSetPriority": "Regular",
                "storageProfile": "ManagedDisks",
                "osDiskSizeGB": 400,
                "acceleratedNetworkingEnabled": true,
                "singlePlacementGroup": false,
                "kubernetesConfig": {
                    "kubeletConfig": {
                        "--register-with-taints": "beta.kubernetes.io/instance-type=Standard_D32s_v3:NoSchedule"
                    }
                }
            }
        ],
        "linuxProfile": {
            "adminUsername": "azureuser",
            "ssh": {
                "publicKeys": [
                    {
                        "keyData": "ssh-rsa XXXX TEST-STG"
                    }
                ]
            }
        }
    }
}
```

1. **Add user Assigned identity** to one of the VMSS  in the cluster (additionally to the cluster managed identity)
	
1. **Update the relevant VMSS with the identities**
	aks-engine update --azure-env AzurePublicCloud --subscription-id <subscription-id> --resource-group <resource group> --location centralus --api-model  ./aksoutput/apimodel.json --client-id <client id> --client-secret <secret>  --node-pool <VMSS name>

**Expected behavior**
Identities should be preserved and not only the cluster default managed identity 

**AKS Engine version**
Version: v0.57.0, GitCommit: e80c40dba, GitTreeState: clean

**Kubernetes version**
"1.16.15"

**Additional context**
Customers will not be able acquire token and fail access azure resources "suddenly".
We are using AAD Pod Identity To manage our identities.

**Comment by jackfrancis** [comment]
Hi @atiasadir 

The part where you "add user assigned identity to one of the VMSS in the cluster" after you build the cluster is what won't work on AKS Engine.

Unfortunately, the AKS Engine toolkit is not resilient to these kinds of out-of-band changes. Once you manually modify/add/delete resources in your aks-engine-created resource group, you should expect weird things to happen.

tl;dr this workflow is not going to work. What will have to happen is for the "additional" user assigned identies to be present in the ARM template that AKS Engine creates. Otherwise the `aks-engine update` operation will deploy an ARM template with no knowledge of that additive stuff, and bad outcomes will happen.

Alternatively, you *may* be able to use `aks-engine addpool` in this workflow. So, rather than "updating" existing VMSS pools, you'd add new, updated pools. Depending on how that user-assigned identity stuff is being applied to the VMSS, that may work. By adding pools, you'll need to manually migrate workloads over to the new pool(s) after they're created, after which you can delete the old ones, and then manually delete them from the AKS Engine api model. See the docs here:

https://github.com/Azure/aks-engine/blob/master/docs/topics/addpool.md

cc @devigned @craiglpeters

**Comment by devigned** [comment]
Hmm... I think [Cluster API Provider Azure (CAPZ)](https://github.com/kubernetes-sigs/cluster-api-provider-azure) would overwrite the user assigned identity as well if a set of identities were provided to the CAPZ infrastructure provider via an [AzureMachinePool](https://capz.sigs.k8s.io/topics/machinepools.html] resource due to resource reconciliation.

I do think that CAPZ would not remove the manually added user assigned identity in the case where the [AzureMachinePool](https://capz.sigs.k8s.io/topics/machinepools.html] resource didn't define a set of user assigned identities based on how the REST payload is interpreted by the VMSS resource provider. I not 100% sure, but I think that would be the behavior. I think that behavior needs to be documented.

**Comment by bridgetkromhout** [comment]
Hello, @atiasadir! Thanks for reaching out. As @jackfrancis explained, AKS Engine does not support this use case. (But CAPZ may suit your needs.)

I'll close this issue since AKS Engine is now deprecated, and we encourage users to choose managed AKS clusters if possible, or to look at CAPZ if they require a self-managed cluster: https://github.com/Azure/aks-engine#project-status

As @devigned mentioned, CAPZ does not preserve manually added user assigned identity. However, @CecileRobertMichon explained to me that this use case could be solved by CAPZ. She said, "CAPZ supports adding multiple user assigned identities to a VM (the first one is used by cloud provider, and beyond that they can add any extra ones they need): [VM Identity - The Cluster API Provider Azure Book (k8s.io)](https://capz.sigs.k8s.io/topics/vm-identity.html#user-assigned)".
